### PR TITLE
42 add typescript

### LINKED ISSUE #42: Add TypeScirpt

### PR DESCRIPTION
**Pull request by ivanbkovacevic**
Added TS to the project.

Maybe tsconfig.json some other settings to be put.

Modelselect.tsx  causing some little problems

### CODE REVIEW

**Comment by itmilos** [comment]
@ivanbkovacevicIt may require some time for me to go through this thoroughly. In the future for similar tasks, we should consider breaking it down into smaller parts.

**Comment by ivanbkovacevic** [comment]
> @ivanbkovacevicIt may require some time for me to go through this thoroughly. In the future for similar tasks, we should consider breaking it down into smaller parts.

@itmilos   yes yes.... Offcourse

**Comment by ivanbkovacevic** [comment]
@itmilos  I extracted types to constant but now Im getting conflicts in some files that was not there yesterday.
App constantly  changes .ts import to .js imports

Help needed

File: src/components/Input.tsx [hunk 1]
```diff
@@ -1,16 +1,33 @@
...
+interface InputProps {
+   type?: string;
+   id: string;
+   help: string;
+   helpText: string;
+   disabled?: boolean;
+   width: number;
+   label: string;
+   value: string | number; 
+   whenChanged: (e:any) => void;
```
Comment (itmilos) [inline src/components/Input.tsx]:
```
whenChanged: (e: React.ChangeEvent<HTMLInputElement>) => void;
}
```

File: src/components/Input.tsx [hunk 2]
```diff
@@ -1,16 +1,33 @@
+import React from 'react';
+import { deviceBrakepoints } from '../config/device-brakepoints'
 import Label from './Label.jsx'
```
Comment (itmilos) [inline src/components/Input.tsx]: remove .jsx

File: src/components/InputGroup.tsx [hunk 3]
```diff
@@ -1,14 +1,22 @@
-import ErrorMessage from './ErrorMessage.jsx'
```
Comment (itmilos) [inline src/components/InputGroup.tsx:1]: remove .jsx

File: src/components/InputGroup.tsx [hunk 4]
```diff
@@ -1,14 +1,22 @@
...
+import React from 'react'
+import ErrorMessage from './ErrorMessage.js'
```
Comment (itmilos) [inline src/components/InputGroup.tsx]: remove .js

File: src/components/InputGroup.tsx [hunk 5]
```diff
@@ -1,14 +1,22 @@
...
 import S from 'styled-components'
 import { useState } from 'react'
+import { InputsProps } from './SplittedInput.js';
```
Comment (itmilos) [inline src/components/InputGroup.tsx]: remove.js

Comment (ivanbkovacevic) [reply src/components/InputGroup.tsx]: @itmilos  I do not know why is that happening. The app is constantly injecting .jsx imports

File: src/components/SavedTemplates.tsx [hunk 6]
```diff
@@ -1,19 +1,25 @@
...
+const SavedTemplates:React.FC<SavedTemplatesProps> = ({useTemplate}) => {
   const[clearallTemplatedMsg, setClearedAllTemplatesMsg] = useState('');
   const[templatedIsUsedMsg, setTemplateIsUsedMsg] = useState('');
 
-  const templates = JSON.parse(localStorage.getItem('templates')) ?? [];
-
-  const handleUseTemplate = (template) => {
+  const pullTemplates = localStorage.getItem('templates');
+  const templates = pullTemplates !== null ? JSON.parse(pullTemplates) : [];
+ console.log(templates);
```
Comment (itmilos) [inline src/components/SavedTemplates.tsx]: remove

File: src/components/Textarea.tsx [hunk 7]
```diff
@@ -1,10 +1,19 @@
...
 import Label from './Label'
 
-const Textarea = ({ label, id, help, helpText, value, whenChanged }) => (
+interface TextareaProps {
+    id: string;
+    help: string;
+    helpText: string;
+    label: string;
+    value: string | number; 
+    whenChanged: (e:any) => void;
```
Comment (itmilos) [inline src/components/Textarea.tsx:11]: `  whenChanged: (e: React.ChangeEvent<HTMLTextAreaElement>) => void;`

File: src/components/SavedTemplates.tsx (hunk 6 above)
Comment (ivanbkovacevic) [reply src/components/SavedTemplates.tsx]: @itmilos   I have made the recommended changes.
Please test the branch before merging into master. It is a  big change and also configuration changes

File: src/pages/Payslip.tsx [hunk 8]
```diff
@@ -959,38 +990,39 @@ function Payslip() {
     const [state, dispatch] = useReducer(reducer, initialState, init)
```
Comment (itmilos) [inline src/pages/Payslip.tsx:187]: `const [state, dispatch] = useReducer<React.Reducer<ReducerState, ReducerAction>>(reducer, initialState, init)
    `

File: src/pages/Payslip.tsx [hunk 9]
```diff
@@ -959,38 +990,39 @@ function Payslip() {
...
-    const onTotalAmountChange = event => dispatch({ type: ACTIONS.TOTAL_AMOUNT, payload: event.target.value })
+    const onPayerChange = (event: { target: { value: any } }) => dispatch({ type: ACTIONS.PAYER_CHANGED, payload: event.target.value })
+    const onPaymentDescriptionChange = (event: { target: { value: any } }) => dispatch({ type: ACTIONS.PAYMENT_DESCRIPTION, payload: event.target.value })
+    const onReceiverChange = (event: { target: { value: any } }) => dispatch({ type: ACTIONS.RECEIVER_CHANGED, payload: event.target.value })
+    const onPayCodeChange = (event: any) => dispatch({ type: ACTIONS.PAYCODE_CHANGED, payload: event })
+    const onCurrencyCode = (event: { target: { value: any } }) => dispatch({ type: ACTIONS.CURRENCY_CHANGED, payload: event.target.value })
+    const onTotalAmountChange = (event: { target: { value: any } }) => dispatch({ type: ACTIONS.TOTAL_AMOUNT, payload: event.target.value })
 
-    const appendZeros = value => {
+    const appendZeros = (value: any) => {
```
Comment (itmilos) [inline src/pages/Payslip.tsx:197]: `   const appendZeros = (value: string) `

not sure please double check

File: src/pages/Payslip.tsx [hunk 10]
```diff
@@ -959,38 +990,39 @@ function Payslip() {
...
         let newValue = value
         while (newValue.length < 13) newValue = '0' + newValue
         dispatch({ type: ACTIONS.ACCOUNT_NUMBER_CHANGE, payload: newValue })
     }
-    const onFixBankNumberChange = event => {
+    const onFixBankNumberChange = (event: { target: { value: string } }) => {
```
Comment (itmilos) [inline src/pages/Payslip.tsx:202]: const onFixBankNumberChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        if (/^[1-9]{0,1}[0-9]{0,2}$/.test(event.target.value))
            dispatch({ type: ACTIONS.BANK_NUMBER_CHANGE, payload: event.target.value })
    }
    const onAccountNumberChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        if (/^[0-9]{0,13}$/.test(event.target.value)) {
            dispatch({ type: ACTIONS.ACCOUNT_NUMBER_CHANGE, payload: event.target.value })
        }
    }
    const onControlNumberChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        if (/^[1-9]{0,1}[0-9]{0,1}$/.test(event.target.value))
            dispatch({ type: ACTIONS.CONTROL_NUMBER_CHANGE, payload: event.target.value })
    }

File: src/pages/Payslip.tsx [hunk 11]
```diff
@@ -959,38 +990,39 @@ function Payslip() {
...
-    const onControlNumberChange = event => {
+    const onControlNumberChange = (event: { target: { value: string } }) => {
         if (/^[1-9]{0,1}[0-9]{0,1}$/.test(event.target.value))
             dispatch({ type: ACTIONS.CONTROL_NUMBER_CHANGE, payload: event.target.value })
     }
 
-    const onSetModelCodeChange = event => dispatch({ type: ACTIONS.MODEL_CODE, payload: event })
-    const onPaymentNumberChange = event => dispatch({ type: ACTIONS.PAYMENT_NUMBER, payload: event.target.value })
+    const onSetModelCodeChange = (event: { value: string; label: string }) => dispatch({ type: ACTIONS.MODEL_CODE, payload: event })
+    const onPaymentNumberChange = (event: { target: { value: any } }) => dispatch({ type: ACTIONS.PAYMENT_NUMBER, payload: event.target.value })
```
Comment (itmilos) [inline src/pages/Payslip.tsx:217]: const onPaymentNumberChange = (event: React.ChangeEvent<HTMLInputElement>)

File: src/pages/Payslip.tsx [hunk 12]
```diff
@@ -959,38 +990,39 @@ function Payslip() {
...
     const resetValues = () => dispatch({ type: ACTIONS.RESET_VALUES })
     
-    const storeTemplate = (templateName) => {
-        const templates = JSON.parse(localStorage.getItem('templates')) ?? [];
+    const storeTemplate = (templateName: any) => {
```
Comment (itmilos) [inline src/pages/Payslip.tsx]: const storeTemplate = (templateName: string)

**Review by itmilos — CHANGES_REQUESTED** [review]
are we using in Payslip:1067
        const accountNumber = params.get('account-number')
        const amount = params.get('amount')

**Pull request merged**